lib/spack/spack/build_systems/python.py

```diff
@@ -160,83 +87,121 @@ def skip_modules(self):
+                shutil.copy2(src, dst)
+                is_script = fs.is_nonsymlink_exe_with_shebang(src)
+                if is_script and not python_is_external:
+                    fs.filter_file(
+                        python_prefix,
+                        os.path.abspath(view.get_projection_for_spec(self.spec)),
+                        dst,
+                    )
+            else:
+                orig_link_target = os.path.realpath(src)
+                new_link_target = os.path.abspath(merge_map[orig_link_target])
+                view.link(new_link_target, dst)
 
-        Requires pip 22.1+, which requires Python 3.7+.
+    def remove_files_from_view(self, view, merge_map):
+        ignore_namespace = False
+        if self.py_namespace:
+            ext_map = view.extensions_layout.extension_map(self.extendee_spec)
+            remaining_namespaces = set(
+                spec.package.py_namespace for name, spec in ext_map.items() if name != self.name
+            )
+            if self.py_namespace in remaining_namespaces:
+                namespace_init = lang.match_predicate(
+                    r"site-packages/{0}/__init__.py".format(self.py_namespace)
+                )
+                ignore_namespace = True
 
-        Args:
-            spec (spack.spec.Spec): build spec
-            prefix (spack.util.prefix.Prefix): installation prefix
+        bin_dir = self.spec.prefix.bin
+        global_view = self.extendee_spec.prefix == view.get_projection_for_spec(self.spec)
 
-        Returns:
-            dict: dictionary of KEY, VALUE settings
-        """
-        return {}
+        to_remove = []
+        for src, dst in merge_map.items():
+            if ignore_namespace and namespace_init(dst):
+                continue
 
-    def install_options(self, spec, prefix):
-        """Extra arguments to be supplied to the setup.py install command.
+            if global_view or not fs.path_contains_subdirectory(src, bin_dir):
+                to_remove.append(dst)
+            else:
+                os.remove(dst)
 
-        Args:
-            spec (spack.spec.Spec): build spec
-            prefix (spack.util.prefix.Prefix): installation prefix
+        view.remove_files(to_remove)
 
-        Returns:
-            list: list of options
-        """
-        return []
 
-    def global_options(self, spec, prefix):
-        """Extra global options to be supplied to the setup.py call before the install
-        or bdist_wheel command.
+class PythonPackage(PythonExtension):
+    """Specialized class for packages that are built using pip."""
 
-        Args:
-            spec (spack.spec.Spec): build spec
-            prefix (spack.util.prefix.Prefix): installation prefix
+    #: Package name, version, and extension on PyPI
+    pypi = None  # type: Optional[str]
 
-        Returns:
-            list: list of options
-        """
-        return []
+    maintainers = ["adamjstewart", "pradyunsg"]
```

Comment:
Is it possible to also add this to `PythonExtension` and/or `PythonPipBuilder`? I would love to be automatically pinged on packages even if they don't subclass from `PythonPackage`

Reply:
Sure. The place to do it is `PythonExtension`. Maybe you're interested in that, but with this PR I think we can code multiple installers for Python packages (e.g. `pip`, `setup.py`, etc.).